--!strict

local constants = {}

export type ColorMapping = { studio: string, vscode: { string } }

constants.ROBLOX_VSCODE_THEME_MAP = {
	{ studio = "Background Color", vscode = { "editor.background" } },
	{ studio = "Selection Color", vscode = { "foreground", "editor.foreground" } },
	{ studio = "Selected Text Color", vscode = { "editor.selectionForeground", "foreground", "editor.foreground" } },
	{
		studio = "Selection Background Color",
		vscode = {
			"editor.selectionBackground",
			"selection.background",
			"editor.inactiveSelectionBackground",
			"editor.selectionHighlightBackground",
			"editor.wordHighlightBackground",
			"editor.findMatchHighlightBackground",
		},
	},
	{
		studio = "Matching Word Background Color",
		vscode = {
			"editor.selectionHighlightBackground",
			"editor.wordHighlightBackground",
			"editor.selectionBackground",
		},
	},
	{
		studio = "Find Selection Background Color",
		vscode = { "editor.findMatchBackground", "editor.findMatchHighlightBackground" },
	},
	{ studio = "Error Color", vscode = { "errorForeground", "editorError.foreground" } },
	{
		studio = "Warning Color",
		vscode = {
			"editorWarning.foreground",
			"editorWarning.border",
			"list.warningForeground",
			"editorOverviewRuler.warningForeground",
			"problemsWarningIcon.foreground",
			"statusBarItem.warningForeground",
			"inputValidation.warningForeground",
			"warningForeground",
			"markup.warning",
			"invalid.deprecated",
			"editorError.foreground",
			"errorForeground",
		},
	},
	{
		studio = "InformationColor",
		vscode = {
			"editorInfo.foreground",
			"editorInfo.border",
			"list.infoForeground",
			"editorOverviewRuler.infoForeground",
			"problemsInfoIcon.foreground",
			"notificationsInfoIcon.foreground",
			"foreground",
			"editor.foreground",
		},
	},
	{
		studio = "HintColor",
		vscode = {
			"editorHint.foreground",
			"editorHint.border",
			"editorOverviewRuler.infoForeground",
			"editorInfo.foreground",
			"foreground",
			"editor.foreground",
		},
	},
	{ studio = "Whitespace Color", vscode = { "editorWhitespace.foreground", "editorLineNumber.activeForeground" } },
	{ studio = "Current Line Highlight Color", vscode = { "editor.lineHighlightBackground", "editor.background" } },
	{
		studio = "Ruler Color",
		vscode = {
			"editorRuler.foreground",
			"editorOverviewRuler.border",
			"editorOverviewRuler.foreground",
			"editor.lineHighlightBorder",
			"editorIndentGuide.background1",
			"editorIndentGuide.background",
			"editorLineNumber.activeForeground",
			"editorLineNumber.foreground",
			"panel.border",
			"editorGroup.border",
		},
	},
	{
		studio = "Bracket Color",
		vscode = {
			"editorBracketHighlight.foreground1",
			"punctuation.table.begin.luau",
			"punctuation.table.end.luau",
			"punctuation.arguments.begin.luau",
			"punctuation.arguments.end.luau",
			"punctuation.definition.parameters.begin.luau",
			"punctuation.definition.parameters.end.luau",
			"punctuation.definition.typeparameters.begin.luau",
			"punctuation.definition.typeparameters.end.luau",
			"meta.brace",
			"#FFD700",
		},
	},
	{
		studio = "Text Color",
		vscode = {
			"variable.other.readwrite.luau",
			"variable.parameter.luau",
			"variable.parameter.function.luau",
			"variable.parameter.function.varargs.luau",
			"variable.other.readwrite",
			"variable.parameter",
			"variable.other",
			"variable",
			"meta.object-literal.key",
			"support",
			"foreground",
			"editor.foreground",
		},
	},
	{
		studio = "Operator Color",
		vscode = {
			"keyword.operator.arithmetic.luau",
			"keyword.operator.assignment.luau",
			"keyword.operator.comparison.luau",
			"keyword.operator.logical.luau",
			"keyword.operator.other.luau",
			"keyword.operator.typecast.luau",
			"keyword.operator.type.luau",
			"keyword.operator.wordlike.luau",
			"keyword.operator",
			"keyword",
		},
	},
	{
		studio = "Number Color",
		vscode = {
			"constant.numeric.binary.luau",
			"constant.numeric.decimal.luau",
			"constant.numeric.hex.luau",
			"keyword.other.unit.luau",
			"constant.numeric",
			"constant",
		},
	},
	{
		studio = "String Color",
		vscode = {
			"string.quoted.double.luau",
			"string.quoted.single.luau",
			"string.interpolated.luau",
			"string.other.multiline.luau",
			"string.quoted",
			"string",
		},
	},
	{
		studio = "Comment Color",
		vscode = {
			"comment.line.double-dash.documentation.luau",
			"comment.line.double-dash.luau",
			"comment.block.luau",
			"comment.line.shebang.luau",
			"comment.luau",
			"comment.line",
			"comment.block",
			"comment",
		},
	},
	{
		studio = "Bool Color",
		vscode = {
			"constant.language.boolean.true.luau",
			"constant.language.boolean.false.luau",
			"constant.language.boolean",
			"constant.language",
			"constant",
		},
	},
	{
		studio = '"nil" Color',
		vscode = { "constant.language.nil.luau", "constant.language.null", "constant.language", "constant" },
	},
	{
		studio = "Function Name Color",
		vscode = {
			"entity.name.function.luau",
			"meta.function.luau",
			"entity.name.function",
			"variable.function",
			"meta.function",
			"support.function",
		},
	},
	{
		studio = '"function" Color',
		vscode = {
			"storage.type.luau",
			"storage.type.function",
			"keyword.declaration.function",
			"keyword.control.luau",
			"keyword",
			"keyword.control",
			"variable.language",
			"meta.function",
		},
	},
	{
		studio = '"local" Color',
		vscode = {
			"storage.modifier.local.luau",
			"storage.modifier.const.luau",
			"storage.modifier.access.luau",
			"storage.modifier.visibility.luau",
			"storage.type.local",
			"storage.modifier",
			"keyword",
		},
	},
	{
		studio = '"self" Color',
		vscode = {
			"variable.language.self.luau",
			"variable.language.metamethod.luau",
			"variable.language.self",
			"variable.instance",
			"variable.language",
			"support",
			"keyword",
			"variable",
		},
	},
	{
		studio = "Luau Keyword Color",
		vscode = {
			"storage.type.luau",
			"storage.type",
			"storage",
			"support.type.primitive.luau",
			"keyword.operator.type.function.luau",
			"keyword.operator.attribute.luau",
			"variable.language",
			"variable",
		},
	},
	{
		studio = "Keyword Color",
		vscode = { "keyword.control.luau", "keyword.control", "keyword" },
	},
	{
		studio = "Built-in Function Color",
		vscode = {
			"support.function.luau",
			"support.constant.luau",
			"support.function",
			"support.constant",
			"entity.name",
			"entity.other",
		},
	},
	{ studio = '"TODO" Color', vscode = { "comment.todo", "keyword.codetag.notation", "variable", "keyword" } },
	{
		studio = "Method Color",
		vscode = {
			"variable.language.metamethod.luau",
			"entity.name.function.luau",
			"entity.name.function.method",
			"entity.name.function",
			"variable.function",
		},
	},
	{
		studio = "Property Color",
		vscode = {
			"variable.other.property.luau",
			"variable.property.luau",
			"variable.other.property",
			"variable.other.member",
			"support.variable.property",
			"variable",
		},
	},
	{
		studio = "TypeColor",
		vscode = {
			"entity.name.type.luau",
			"support.type.primitive.luau",
			"entity.name.type",
			"support.type",
			"storage.type",
			"support.class",
			"entity.name.class",
		},
	},
	{
		studio = "Debugger Current Line Color",
		vscode = {
			"editor.stackFrameHighlightBackground",
			"editor.focusedStackFrameHighlightBackground",
			"editor.lineHighlightBackground",
			"editor.findMatchBackground",
			"editor.findMatchHighlightBackground",
			"debugToolBar.background",
			"statusBar.debuggingBackground",
			"editorInfo.foreground",
		},
	},
	{
		studio = "Debugger Error Line Color",
		vscode = {
			"editor.stackFrameErrorHighlightBackground",
			"editorError.background",
			"editorError.foreground",
			"editor.findMatchBackground",
		},
	},
	{
		studio = "IndentationRulerColor",
		vscode = {
			"editorIndentGuide.background1",
			"editorIndentGuide.background",
			"editorIndentGuide.activeBackground",
		},
	},
	{
		studio = "Script Editor Scrollbar Background Color",
		vscode = { "scrollbar.shadow", "editorOverviewRuler.background", "editor.background" },
	},
	{
		studio = "Script Editor Scrollbar Handle Color",
		vscode = {
			"scrollbarSlider.background",
			"scrollbarSlider.activeBackground",
			"scrollbarSlider.hoverBackground",
			"editorOverviewRuler.background",
			"editorOverviewRuler.border",
			"panel.border",
			"editorGroup.border",
			"widget.border",
			"tab.border",
			"editorWidget.border",
		},
	},
	{
		studio = "Primary Text Color",
		vscode = { "foreground", "editor.foreground", "editorSuggestWidget.foreground", "menu.foreground" },
	},
	{
		studio = "Secondary Text Color",
		vscode = {
			"descriptionForeground",
			"editorSuggestWidget.descriptionForeground",
			"foreground",
			"editor.foreground",
		},
	},
	{
		studio = "Active Color",
		vscode = { "focusBorder", "button.background", "list.activeSelectionBackground", "tab.activeBorderTop" },
	},
	{
		studio = "Active Hover Over Color",
		vscode = { "list.activeSelectionForeground", "list.hoverForeground", "foreground" },
	},
	{
		studio = "Hover Over Color",
		vscode = { "list.hoverBackground", "editorHoverWidget.background", "editorWidget.background" },
	},
	{
		studio = "Select Color",
		vscode = { "list.activeSelectionBackground", "editor.selectionBackground", "menu.selectionBackground" },
	},
	{
		studio = "Select/Hover Color",
		vscode = {
			"list.hoverBackground",
			"list.inactiveSelectionBackground",
			"editor.selectionBackground",
			"editor.inactiveSelectionBackground",
			"menu.selectionBackground",
			"editor.selectionHighlightBackground",
			"editorSuggestWidget.selectedBackground",
		},
	},
	{
		studio = "Menu Item Background Color",
		vscode = {
			"menu.background",
			"editorSuggestWidget.background",
			"dropdown.background",
			"editorWidget.background",
		},
	},
	{
		studio = "Selected Menu Item Background Color",
		vscode = {
			"menu.selectionBackground",
			"editorSuggestWidget.selectedBackground",
			"list.activeSelectionBackground",
			"editor.selectionBackground",
		},
	},
} :: { ColorMapping }

return constants
